modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1282 @@
+      var totalTimeout = timeout.tv_sec*1000000 + timeout.tv_usec;
+      if totalTimeout >= t.elapsed(TimeUnits.microseconds) {
+        const remainingMicroSeconds = ((totalTimeout - elapsedTime)%1000000);
+        const remainingSeconds = ((totalTimeout - elapsedTime)/1000000);
+        return this.send(data, address, new timeval(remainingSeconds, remainingMicroSeconds));
+      }
+      throw SystemError.fromSyserr(err_out, "send failed");
+    }
+    return this.send(data, address, timeout);
+  }
+  return length;
+}
+
+inline operator !=(in lhs: udpSocket,in rhs: udpSocket) {
+  return lhs.socketFd != rhs.socketFd;
+}
+
+proc udpSocket.writeThis(f) throws {
+  f.write("(","addr:",this.addr,",fd:",this.socketFd);
+}
+
+pragma "no doc"
+proc setSockOpt(socketFd: fd_t, level: c_int, optname: c_int, ref value: c_int) throws {
+  var optlen = sizeof(value):socklen_t;
+  var ptroptval = c_ptrTo(value);
+  var err_out = sys_setsockopt(socketFd, level, optname, ptroptval:c_void_ptr, optlen);
+  if err_out != 0 {
+    throw SystemError.fromSyserr(err_out, "Failed to set socket option");
+  }
+}
+
+/*
+  Set the value of the given socket option (see `setsockopt(2) </https://pubs.opengroup.org/onlinepubs/9699919799/functions/setsockopt.html#>`_)
+  on provided :type:`tcpConn`. The needed symbolic constants (SO_* etc.)
+  are defined in the :mod:`Sys`.
+
+  .. code-block:: Chapel
+
+    setSockOpt(socket, IPPROTO_TCP, TCP_QUICKACK, 1:c_int);
+
+  :arg socket: socket to set option on
+  :type socket: `tcpConn`
+  :arg level: protocol level
+  :type level: `int(32)`
+  :arg optname: option to set.
+  :type optname: `int(32)`
+  :arg value: value to set on option
+  :type value: `int(32)`
+  :throws SystemError: Upon incompatible arguments
+                        and socket.
+*/
+proc setSockOpt(ref socket: tcpConn, level: c_int, optname: c_int, value: c_int) throws {
+  var socketFd = socket.socketFd;
+  setSockOpt(socketFd, level, optname, value);
+}
+
+/*
+  See above
+*/
+proc setSockOpt(ref socket: tcpListener, level: c_int, optname: c_int, value: c_int) throws {
+  var socketFd = socket.socketFd;
+  setSockOpt(socketFd, level, optname, value);
+}
+
+/*
+  See above
+*/
+proc setSockOpt(ref socket: udpSocket, level: c_int, optname: c_int, value: c_int) throws {
+  var socketFd = socket.socketFd;
+  setSockOpt(socketFd, level, optname, value);
+}
+
+pragma "no doc"
+proc setSockOpt(socketFd:fd_t, level: c_int, optname: c_int, ref value: bytes) throws {
+  var optlen = value.size:socklen_t;
+  var ptroptval = value.c_str();
+  var err_out = sys_setsockopt(socketFd, level, optname, ptroptval, optlen);
+  if err_out != 0 {
+    throw SystemError.fromSyserr(err_out, "Failed to set socket option");
+  }
+}
+
+/*
+  Overload for :proc:`setSockOpt` that allows setting a :mod:`bytes <Bytes>` value
+  on socket option. It is up to the caller to ensure that the
+  `value` which is a :type::mod:`bytes <Bytes>` parameter contains the proper bits.
```

Comment:
python has three overloads for setSockOpt so i transferred all three of them

Reply:
Sure. It would still be good to show an example that uses it in the docs.

Reply:
In python they have capibility of writing c structs as bytestrings so if we have it in chapel then only this procedure will be useful for us

Reply:
Sure. But can you show an example for a setsockopt call that would use such a C structure that's more complex than an int or pointer?

Reply:
@king-11 you don't have to be able to show the example in Chapel. I am just interested in an example of when this overload would be used, even if it would be hard or awkward to use in Chapel today. I.e. can you name a setsockopt call that uses such a struct encoded as bytes? That we can learn more about with the C documentation?

Reply:
@mppf as per the `setsockopt` there are three options that need to use C structs:

- **SO_LINGER** uses `linger` struct
- **SO_RCVTIMEO** uses `timeval` struct
- **SO_SNDTIMEO** uses `timeval` struct

Reply:
Sure. Can you just mention these specifically in the docs comment? E.g.

```
  Overload for :proc:`setSockOpt` that allows setting a :mod:`bytes <Bytes>` value
  on socket option.
  This is useful for `setsockopt` calls that work with a C struct, including `SO_LINGER`, `SO_RCVTIMEO`, and `SO_SNDTIMEO`.
  It is up to the caller to ensure that the
  `value` which is a :type::mod:`bytes <Bytes>` parameter contains the proper bits.